test/unit_tests/src/test_timestamp_gen.cpp

```diff
@@ -18,31 +18,114 @@
 #   define BOOST_TEST_MODULE cassandra
 #endif
 
+#include "get_time.hpp"
+#include "logger.hpp"
 #include "timestamp_generator.hpp"
 
 #include <boost/test/unit_test.hpp>
 
+static void clock_skew_log_callback(const CassLogMessage* message, void* data) {
+  std::string msg(message->message);
+  int* counter = reinterpret_cast<int*>(data);
+  if (msg.find("Clock skew detected") != std::string::npos) {
+    (*counter)++;
+  }
+}
 
 BOOST_AUTO_TEST_SUITE(timestamp_gen)
 
 BOOST_AUTO_TEST_CASE(server)
 {
- cass::ServerSideTimestampGenerator gen;
- BOOST_CHECK(gen.next() == CASS_INT64_MIN);
+  cass::ServerSideTimestampGenerator gen;
+  BOOST_CHECK(gen.next() == CASS_INT64_MIN);
 }
 
- // This test may log: "Sub-millisecond counter overflowed, some query timestamps will not be distinct"
 BOOST_AUTO_TEST_CASE(monotonic)
 {
- cass::MonotonicTimestampGenerator gen;
- // This should get at least 1000 unique timestamps
- int64_t ts = gen.next();
- BOOST_CHECK(gen.next() > CASS_INT64_MIN);
- for (int i = 0; i < 999; ++i) {
-   int64_t now = gen.next();
-   BOOST_CHECK(now > ts);
- }
- BOOST_CHECK(gen.next() >= ts);
+  const int NUM_TIMESTAMPS = 5000000;
+
+  cass::MonotonicTimestampGenerator gen;
+  int counter = 0;
+
+  cass::Logger::set_callback(clock_skew_log_callback, &counter);
+
+  uint64_t start = cass::get_time_since_epoch_ms();
+
+  int64_t prev = gen.next();
+  for (int i = 0; i < NUM_TIMESTAMPS; ++i) {
+    int64_t now = gen.next();
+    // Verify that timestamps are alway increasing
+    BOOST_CHECK(now > prev);
+    prev = now;
+  }
+  double elapsed = (cass::get_time_since_epoch_ms() - start) / 1000.0; // Convert to seconds
+
+  // We can generate at most 1,000,000 timestamps in a second. If we exceed this
+  // limit and the clock skew threshold then a log should have been printed.
+  double timestamp_rate = static_cast<double>(NUM_TIMESTAMPS) / elapsed;
+  if (timestamp_rate > 1000000.0 && elapsed > 1.0) {
+    BOOST_CHECK(counter > 0);
```

Comment:
In practice, does the test get here? Does it take at least 1.0 seconds to get 5,000,000 timestamps?